c++/src/kj/memory.h

```diff
@@ -168,6 +170,46 @@ class NullDisposer: public Disposer {
   void disposeImpl(void* pointer) const override {}
 };
 
+
+// =======================================================================================
+// Ptr Counters
+
+#ifdef KJ_DEBUG
+#define KJ_ASSERT_PTR_COUNTERS
+// When defined, keeps track of active Ptr<T> instances and asserts validity of their ownership
+#endif
+
+namespace _ {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+class AtomicPtrCounter {
+// AtomicPtrCounter uses atomic operations to keep track of active pointers.
+
+public:
+  void dec() {
+    size_t c = count.load(std::memory_order_consume);
+    KJ_IREQUIRE(c > 0, "unbalanced dec");
+    count.fetch_sub(1, std::memory_order_release);
```

Comment:
I think there's a race condition here. if multiple threads call dec at once, there's a possibility of both observing c == 1 then both subtracting resulting in a missed unbalanced dec (c == -1).

I think it would be more correct to capture the return value of fetch_sub then assert it's nonnegative

Reply:
Actually, maybe the memory ordering makes things okay 🫣

Reply:
I think it is better to change the count variable to `int64_t` and then this would be something like:
```cpp
int64_t c = count.fetch_sub(1, std::memory_order_consume);
KJ_IREQUIRE(c == 0, "unbalanced dec");
```

Reply:
> I think it would be more correct to capture the return value of fetch_sub then assert it's nonnegative

I agree with this.

> I think it is better to change the count variable to int64_t and then this would be something like:

Hello Aditya, I'm not sure if we've met.

I don't think we need to make the count signed. It shouldn't be possible for the count to go negative in the first place -- if it can, that already suggest there is undefined behavior. I think we should probably make this function abort the process (not just throw) if it sees the counter wrap around. Then it doesn't matter if we correctly handle the case that the counter was already below zero, because the process will crash immediately when the counter hits zero.

Reply:
Hello Kenton, I don't believe we have met either. 

I do see your point, if the goal is to crash the system immediately. 
The other concern I have is counting up to the size of an unsigned 32 bit integer and overflowing it is possible with single increments on multiple threads in about 30 minutes. I'm not sure how out of scope it is that this can occur, that is another reason to suggest using a 64 bit width integer for this, even if it should be a `uint64_t`. Unfortunately, I'm not super familiar with this system, so I'm not 100% sure what the context is, @garrettgu10 mentioned it in passing and I thought I'd look into the race.

Reply:
The counter is counting the number of pointers pointing at an object. As a size_t it cannot possibly overflow, since the pointers required to overflow it would not fit into memory.

Reply:
Done. I also wrote some notes regarding the use of release/consume memory ordering. I'm not the expert here, is my understanding correct?

I am also not sure if it is guaranteed that you will ever observe `0` previous value when dec/inc are unbalanced.
At this point I think I'd rather dropped the assert: it is not something that could be really triggered by the user (but Ptr implementer).

Reply:
> I am also not sure if it is guaranteed that you will ever observe 0 previous value when dec/inc are unbalanced.

It is guaranteed, there's no way you can skip a value if you're using the result of `fetch_sub()`.

> At this point I think I'd rather dropped the assert:

I'd be OK with that, but what you have currently looks right.